File: packages/gheim-py/src/gheim/openai/_completions.py
"""Legacy completions API — single ``prompt`` field in, ``choices[*].text`` out.

Pre-tokenized prompts (``list[int]``) are passed through unchanged.
"""
from __future__ import annotations

from dataclasses import dataclass
from typing import Annotated, Any

from ..core.session import Session
from ..core.stream import StreamDeanonymizer
from ..detectors import Detector
from ._base import (
    AsyncStreamingResponse,
    StreamingResponse,
    anonymize_value,
    resolve_call_session,
)


def _anonymize_completions_kwargs(
    kwargs: dict[str, Any], session: Session, detector: Detector
) -> None:
    if "prompt" in kwargs:
        kwargs["prompt"] = anonymize_value(kwargs["prompt"], session, detector)


def _restore_completions_response(response: Any, session: Session) -> None:
    try:
        for choice in response.choices:
            text = getattr(choice, "text", None)
            if isinstance(text, str):
                choice.text = session.restore(text)
    except AttributeError:
        pass


def _patch_completions_chunk(chunk: Any, buf: StreamDeanonymizer) -> None:
    choices = getattr(chunk, "choices", None) or []
    for choice in choices:
        text = getattr(choice, "text", None)
        if isinstance(text, str):
            choice.text = buf.feed(text)


def _make_completions_tail_chunk(template: Any, tail: str) -> Any:
    try:
        cloned = template.model_copy(deep=True)
        if cloned.choices:
            cloned.choices[0].text = tail
        return cloned
    except Exception:  # pragma: no cover — pure fallback path
        return None


@dataclass
class CompletionsProxy:
    inner: Annotated[Any, "openai client's completions object (legacy)."]
    detector: Annotated[Detector, "Default detector for this client."]

    def create(self, *args: Any, **kwargs: Any) -> Any:
        sess, detector = resolve_call_session(kwargs, self.detector)
        _anonymize_completions_kwargs(kwargs, sess, detector)
        stream = bool(kwargs.get("stream"))
        response = self.inner.create(*args, **kwargs)
        if stream:
            return StreamingResponse(
                stream=response,
                session=sess,
                mutate_chunk=_patch_completions_chunk,
                make_tail_chunk=_make_completions_tail_chunk,
            )
        _restore_completions_response(response, sess)
        return response


@dataclass
class AsyncCompletionsProxy:
    inner: Annotated[Any, "openai async client's completions object."]
    detector: Annotated[Detector, "Default detector for this client."]

    async def create(self, *args: Any, **kwargs: Any) -> Any:
        sess, detector = resolve_call_session(kwargs, self.detector)
        _anonymize_completions_kwargs(kwargs, sess, detector)
        stream = bool(kwargs.get("stream"))
        response = await self.inner.create(*args, **kwargs)
        if stream:
            return AsyncStreamingResponse(
                stream=response,
                session=sess,
                mutate_chunk=_patch_completions_chunk,
                make_tail_chunk=_make_completions_tail_chunk,
            )
        _restore_completions_response(response, sess)
        return response

File: packages/gheim-js/src/openai/_completions.ts
/** Legacy completions API — text-in / text-out. */
import type OpenAIClass from "openai";
import { Session } from "../core/session.ts";
import { StreamDeanonymizer } from "../core/stream.ts";
import type { Detector } from "../detectors/base.ts";
import { StreamingResponse, anonymizeValue, type CallExtras, resolveCallSession } from "./_base.ts";

interface RawCompletions {
  create(body: unknown): Promise<unknown>;
}

interface LegacyChunk {
  choices: Array<{ text?: string | null; index: number; finish_reason?: string | null }>;
  id: string;
  model: string;
  object?: string;
  created?: number;
}

function patchLegacyChunk(chunk: LegacyChunk, buf: StreamDeanonymizer): void {
  for (const choice of chunk.choices ?? []) {
    if (typeof choice.text === "string") {
      choice.text = buf.feed(choice.text);
    }
  }
}

function makeLegacyTail(template: LegacyChunk, tail: string): LegacyChunk {
  return {
    id: template.id,
    object: template.object ?? "text_completion.chunk",
    created: template.created ?? 0,
    model: template.model,
    choices: [{ index: 0, text: tail, finish_reason: null }],
  };
}

export class GheimCompletions {
  constructor(
    private readonly clientPromise: Promise<OpenAIClass>,
    private readonly defaultDet: Detector,
  ) {}

  async create(body: Record<string, unknown> & CallExtras): Promise<unknown> {
    const { session, detector, rest } = resolveCallSession(body, this.defaultDet);
    if ("prompt" in rest) {
      rest.prompt = await anonymizeValue(rest.prompt, session, detector);
    }
    const client = await this.clientPromise;
    const response = await (client as unknown as { completions: RawCompletions }).completions.create(rest);
    if (rest.stream) {
      return new StreamingResponse(
        response as AsyncIterable<LegacyChunk>,
        session,
        patchLegacyChunk,
        makeLegacyTail,
      );
    }
    const r = response as { choices?: Array<{ text?: string | null }> };
    for (const choice of r.choices ?? []) {
      if (typeof choice.text === "string") {
        choice.text = session.restore(choice.text);
      }
    }
    return response;
  }
}
